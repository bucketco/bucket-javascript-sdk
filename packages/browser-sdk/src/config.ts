import { version } from "../package.json";

export const API_BASE_URL = "https://front.bucket.co";
export const APP_BASE_URL = "https://app.bucket.co";
export const SSE_REALTIME_BASE_URL = "https://livemessaging.bucket.co";

export const SDK_VERSION_HEADER_NAME = "reflag-sdk-version";

export const SDK_VERSION = `browser-sdk/${version}`;
export const FEATURE_EVENTS_PER_MIN = 1;
