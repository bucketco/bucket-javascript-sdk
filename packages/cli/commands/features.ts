import chalk from "chalk";
import { Command } from "commander";
import { outputFile } from "fs-extra/esm";

import {
  createFeature,
  genFeatureTypes,
  listFeatures,
  rolloutFeature,
} from "../services/features.js";
import { checkAuth } from "../utils/auth.js";
import { GEN_TYPES_FILE } from "../utils/constants.js";

export const featuresCommand = new Command("features").description(
  "Manage features",
);

featuresCommand
  .command("list")
  .description("List all features")
  .requiredOption("-a, --appId <appId>", "Get all features in app")
  .action(async (options) => {
    checkAuth();
    try {
      const features = await listFeatures(options.appId);
      console.log(chalk.green("Features:"));
      console.table(features);
    } catch (error) {
      if (error instanceof Error) {
        console.error(chalk.red("Error fetching features:", error.message));
      }
    }
  });

featuresCommand
  .command("create")
  .description("Create a new feature")
  .action(async () => {
    checkAuth();
    try {
      await createFeature();
    } catch (error) {
      if (error instanceof Error) {
        console.error(chalk.red("Error creating feature:", error.message));
      }
    }
  });

featuresCommand
  .command("types")
  .description("Generate feature types")
  .requiredOption("-a, --appId <appId>", "Get all features in app")
  .action(async (options) => {
    checkAuth();
    try {
      const types = await genFeatureTypes(options.appId);
      await outputFile(GEN_TYPES_FILE, types);
    } catch (error) {
      if (error instanceof Error) {
        console.error(chalk.red("Error listing feature types:", error.message));
      }
    }
  });

featuresCommand
  .command("rollout")
  .description("Rollout a feature")
  .action(async () => {
    checkAuth();
    try {
      await rolloutFeature();
    } catch (error) {
      if (error instanceof Error) {
        console.error(chalk.red("Error rolling out feature:", error.message));
      }
    }
  });
