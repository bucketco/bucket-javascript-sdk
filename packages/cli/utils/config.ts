import { readJson, writeJson } from "fs-extra/esm";
import { findUp } from "find-up";

import { REPO_CONFIG_FILE } from "./constants.js";
import { Datatype } from "./gen.js";

export const generatedPackageName = "_bucket";

type Config = {
  features: ConfigFeatureDefs;
  codeGenBasePath?: string;
};

export type ConfigFeatureDef = {
  key: string;
  access?: boolean;
  config?: Datatype;
};

export type ConfigFeatureDefs = Array<string | ConfigFeatureDef>;

let config: Config = {
  features: [],
  codeGenBasePath: "node_modules/",
};

/**
 * Instantly return a specified key's value or the entire config object.
 */
export function getConfig(): Config;
export function getConfig(key?: keyof Config) {
  return key ? config[key] : config;
}

export async function findRepoConfig() {
  return await findUp(REPO_CONFIG_FILE);
}

/**
 * Read the config file and return either a specified key's value or the entire config object.
 */
export async function readConfigFile() {
  const location = await findRepoConfig();
  if (!location) {
    throw new Error("No bucket.config.js file found.");
  }
  return await readJson(location);
}

/**
 * Write a new value to the config file.
 */
export async function writeConfigFile(config: object, location?: string) {
  const path = location ? location : await findRepoConfig();
  if (!path) throw new Error("writeConfigFile: Could not find config file.");
  await writeJson(path, config, { spaces: 2 });
}

export async function loadConfig() {
  config = await readConfigFile();
}

export async function configFileExists() {
  return !!(await findRepoConfig());
}
