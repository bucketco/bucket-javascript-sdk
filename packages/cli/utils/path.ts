export function stripTrailingSlash<T extends string | undefined>(str: T): T {
  return str?.endsWith("/") ? (str.slice(0, -1) as T) : str;
}
