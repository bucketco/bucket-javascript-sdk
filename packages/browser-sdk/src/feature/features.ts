import { FEATURE_EVENTS_PER_MIN } from "../config";
import { HttpClient } from "../httpClient";
import { Logger, loggerWithPrefix } from "../logger";
import RateLimiter from "../rateLimiter";

import {
  FeatureCache,
  isObject,
  parseAPIFeaturesResponse,
} from "./featureCache";
import maskedProxy from "./maskedProxy";

export type APIFeatureResponse = {
  key: string;
  isEnabled: boolean;
  targetingVersion?: number;
};

export type APIFeaturesResponse = Record<string, APIFeatureResponse>;

export type Features = Record<string, boolean>;

export type FeaturesOptions = {
  fallbackFeatures?: string[];
  timeoutMs?: number;
  staleWhileRevalidate?: boolean;
};

type Config = {
  fallbackFeatures: string[];
  timeoutMs: number;
  staleWhileRevalidate: boolean;
};

export const DEFAULT_FEATURES_CONFIG: Config = {
  fallbackFeatures: [],
  timeoutMs: 5000,
  staleWhileRevalidate: false,
};

// Deep merge two objects.
export function mergeDeep(
  target: Record<string, any>,
  ...sources: Record<string, any>[]
): Record<string, any> {
  if (!sources.length) return target;
  const source = sources.shift();

  if (isObject(target) && isObject(source)) {
    for (const key in source) {
      if (isObject(source[key])) {
        if (!target[key]) Object.assign(target, { [key]: {} });
        mergeDeep(target[key], source[key]);
      } else {
        Object.assign(target, { [key]: source[key] });
      }
    }
  }

  return mergeDeep(target, ...sources);
}

export type FeaturesResponse = {
  success: boolean;
  features: APIFeaturesResponse;
};

export function validateFeaturesResponse(response: any) {
  if (!isObject(response)) {
    return;
  }

  if (typeof response.success !== "boolean" || !isObject(response.features)) {
    return;
  }
  const features = parseAPIFeaturesResponse(response.features);
  if (!features) {
    return;
  }

  return {
    success: response.success,
    features,
  };
}

export function flattenJSON(obj: Record<string, any>): Record<string, any> {
  const result: Record<string, any> = {};
  for (const key in obj) {
    if (typeof obj[key] === "object") {
      const flat = flattenJSON(obj[key]);
      for (const flatKey in flat) {
        result[`${key}.${flatKey}`] = flat[flatKey];
      }
    } else if (typeof obj[key] === "undefined") {
      continue;
    } else {
      result[key] = obj[key];
    }
  }
  return result;
}

export function clearFeatureCache() {
  localStorage.clear();
}

export const FEATURES_STALE_MS = 60000; // turn stale after 60 seconds, optionally reevaluate in the background
export const FEATURES_EXPIRE_MS = 30 * 24 * 60 * 60 * 1000; // expire entirely after 30 days

const localStorageCacheKey = `__bucket_features`;

export class FeaturesClient {
  private cache: FeatureCache;
  private features: Features | undefined;
  private config: Config;
  private rateLimiter: RateLimiter;
  private logger: Logger;

  constructor(
    private httpClient: HttpClient,
    private context: {
      user?: Record<string, any>;
      company?: Record<string, any>;
      other?: Record<string, any>;
    },
    logger: Logger,
    options?: FeaturesOptions & {
      cache?: FeatureCache;
      rateLimiter?: RateLimiter;
    },
  ) {
    this.logger = loggerWithPrefix(logger, "[Features]");
    this.cache = options?.cache
      ? options.cache
      : new FeatureCache({
          storage: {
            get: () => localStorage.getItem(localStorageCacheKey),
            set: (value) => localStorage.setItem(localStorageCacheKey, value),
          },
          staleTimeMs: FEATURES_STALE_MS,
          expireTimeMs: FEATURES_EXPIRE_MS,
        });
    this.config = { ...DEFAULT_FEATURES_CONFIG, ...options };
    this.rateLimiter =
      options?.rateLimiter ??
      new RateLimiter(FEATURE_EVENTS_PER_MIN, this.logger);
  }

  async initialize() {
    const features = (await this.maybeFetchFeatures()) || {};
    this.setFeatures(features);
  }

  private setFeatures(features: APIFeaturesResponse) {
    const proxiedFeatures = maskedProxy(features, (fs, key) => {
      this.sendCheckEvent({
        key,
        version: features[key]?.targetingVersion,
        value: features[key]?.isEnabled ?? false,
      }).catch((e) => {
        this.logger.error("error sending feature check event", e);
      });
      return fs[key]?.isEnabled || false;
    });
    this.features = proxiedFeatures;
  }

  getFeatures(): Features | undefined {
    return this.features;
  }

  private fetchParams() {
    const flattenedContext = flattenJSON({ context: this.context });
    const params = new URLSearchParams(flattenedContext);
    // publishableKey should be part of the cache key
    params.append("publishableKey", this.httpClient.publishableKey);

    // sort the params to ensure that the URL is the same for the same context
    params.sort();

    return params;
  }

  private async maybeFetchFeatures(): Promise<APIFeaturesResponse | undefined> {
    const cacheKey = this.fetchParams().toString();
    const cachedItem = this.cache.get(cacheKey);

    if (cachedItem) {
      if (!cachedItem.stale) return cachedItem.features;

      // serve successful stale cache if `staleWhileRevalidate` is enabled
      if (this.config.staleWhileRevalidate) {
        // re-fetch in the background, but immediately return last successful value
        this.fetchFeatures()
          .then((features) => {
            if (!features) return;

            this.cache.set(cacheKey, {
              features,
            });
            this.setFeatures(features);
          })
          .catch(() => {
            // we don't care about the result, we just want to re-fetch
          });
        return cachedItem.features;
      }
    }

    // if there's no cached item or there is a stale one but `staleWhileRevalidate` is disabled
    // try fetching a new one
    const fetchedFeatures = await this.fetchFeatures();

    if (fetchedFeatures) {
      this.cache.set(cacheKey, {
        features: fetchedFeatures,
      });

      return fetchedFeatures;
    }

    if (cachedItem) {
      // fetch failed, return stale cache
      return cachedItem.features;
    }

    // fetch failed, nothing cached => return fallbacks
    return this.config.fallbackFeatures.reduce((acc, key) => {
      acc[key] = {
        key,
        isEnabled: true,
      };
      return acc;
    }, {} as APIFeaturesResponse);
  }

  public async fetchFeatures(): Promise<APIFeaturesResponse | undefined> {
    const params = this.fetchParams();
    try {
      const res = await this.httpClient.get({
        path: "/features/enabled",
        timeoutMs: this.config.timeoutMs,
        params,
      });

      if (!res.ok) {
        let errorBody = null;
        try {
          errorBody = await res.json();
        } catch (e) {
          // ignore
        }

        throw new Error(
          "unexpected response code: " +
            res.status +
            " - " +
            JSON.stringify(errorBody),
        );
      }
      const typeRes = validateFeaturesResponse(await res.json());
      if (!typeRes || !typeRes.success) {
        throw new Error("unable to validate response");
      }

      return typeRes.features;
    } catch (e) {
      this.logger.error("error fetching features: ", e);
      return;
    }
  }

  /**
   * Send a feature "check" event.
   *
   *
   * @param feature - The feature to send the event for.
   */
  async sendCheckEvent(feature: {
    key: string;
    value: boolean;
    version?: number;
  }) {
    const rateLimitKey = `${this.fetchParams().toString()}:${feature.key}:${feature.version}:${feature.value}`;

    await this.rateLimiter.rateLimited(rateLimitKey, async () => {
      const payload = {
        action: "check",
        key: feature.key,
        targetingVersion: feature.version,
        evalContext: this.context,
        evalResult: feature.value,
      };

      this.httpClient
        .post({
          path: "features/events",
          body: payload,
        })
        .catch((e: any) => {
          this.logger.warn(`failed to send feature check event`, e);
        });

      this.logger.debug(`sent feature event`, payload);
    });

    return feature.value;
  }
}
