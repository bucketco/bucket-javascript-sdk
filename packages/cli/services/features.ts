import {
  ConfigFeatureDefs,
  generatedPackageName,
  readConfigFile,
  writeConfigFile,
} from "../utils/config.js";
import { FeatureDef, genDTS, genJs } from "../utils/gen.js";
import { outputFile } from "fs-extra";
import path from "path";

const packageJson = {
  name: generatedPackageName,
  version: "1.0.0",
  description: "Autogenerated package for typed features for Bucket",
  main: "index.js",
  types: "index.d.ts",
  exports: {
    ".": {
      import: "./index.js",
      types: "./index.d.ts",
    },
  },
};

export async function genFeatureTypes(
  configFeatures: ConfigFeatureDefs,
  basePath: string,
) {
  const features = configFeatures.map((feature) => ({
    key: typeof feature === "string" ? feature : feature.key,
    access: typeof feature === "string" ? true : (feature.access ?? true),
    config: typeof feature === "string" ? undefined : feature.config,
  }));

  await outputFile(
    path.join(basePath, generatedPackageName, `index.d.ts`),
    genDTS(features),
  );
  await outputFile(
    path.join(basePath, generatedPackageName, `index.js`),
    genJs(features),
  );
  await outputFile(
    path.join(basePath, generatedPackageName, `package.json`),
    JSON.stringify(packageJson, null, 2),
  );
}

export async function addFeatureToConfig(feature: FeatureDef) {
  const config = await readConfigFile();

  if (feature.access && feature.config === undefined)
    config.features.push(feature.key);
  else config.features.push(feature);

  await writeConfigFile(config);
}
