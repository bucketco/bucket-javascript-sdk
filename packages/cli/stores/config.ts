import { Ajv, ValidateFunction } from "ajv";
import { findUp } from "find-up";
import JSON5 from "json5";
import { readFile, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import {
  CONFIG_FILE_NAME,
  DEFAULT_API_URL,
  DEFAULT_BASE_URL,
  DEFAULT_TYPES_OUTPUT,
  SCHEMA_URL,
} from "../utils/constants.js";
import { ConfigValidationError, handleError } from "../utils/errors.js";
import { stripTrailingSlash } from "../utils/path.js";

export const keyFormats = [
  "custom",
  "pascalCase",
  "camelCase",
  "snakeCaseUpper",
  "snakeCaseLower",
  "kebabCaseUpper",
  "kebabCaseLower",
] as const;

export type KeyFormat = (typeof keyFormats)[number];

type Config = {
  $schema: string;
  baseUrl: string;
  apiUrl: string;
  appId: string | undefined;
  typesOutput: string;
  keyFormat: KeyFormat;
};

const defaultConfig: Config = {
  $schema: SCHEMA_URL,
  baseUrl: DEFAULT_BASE_URL,
  apiUrl: DEFAULT_API_URL,
  appId: undefined,
  typesOutput: DEFAULT_TYPES_OUTPUT,
  keyFormat: "custom",
};

class ConfigStore {
  protected config: Config = { ...defaultConfig };
  protected configPath: string | undefined;
  protected projectPath: string | undefined;
  protected validateConfig: ValidateFunction | undefined;

  async initialize() {
    await this.createValidator();
    await this.loadConfigFile();
  }

  protected async createValidator() {
    try {
      // Using current config store file, resolve the schema.json path
      const filePath = fileURLToPath(import.meta.url);
      const schemaPath = join(
        filePath.substring(0, filePath.indexOf("cli") + 3),
        "schema.json",
      );
      const content = await readFile(schemaPath, "utf-8");
      const parsed = JSON5.parse<Config>(content);
      const ajv = new Ajv();
      this.validateConfig = ajv.compile(parsed);
    } catch {
      void handleError(new Error("Failed to load the config schema"), "Config");
    }
  }

  protected async loadConfigFile() {
    if (!this.validateConfig) {
      void handleError(new Error("Failed to load the config schema"), "Config");
      return;
    }

    try {
      const packageJSONPath = await findUp("package.json");
      this.configPath = await findUp(CONFIG_FILE_NAME);
      this.projectPath = dirname(
        this.configPath ?? packageJSONPath ?? process.cwd(),
      );

      if (!this.configPath) return;

      const content = await readFile(this.configPath, "utf-8");
      const parsed = JSON5.parse<Partial<Config>>(content);
      parsed.baseUrl = stripTrailingSlash(parsed.baseUrl?.trim());
      parsed.apiUrl = stripTrailingSlash(parsed.apiUrl?.trim());

      if (!this.validateConfig!(parsed)) {
        void handleError(
          new ConfigValidationError(this.validateConfig!.errors),
          "Config",
        );
      }

      this.config = { ...this.config, ...parsed };
    } catch {
      // No config file found
    }
  }

  /**
   * Create a new config file with initial values.
   * @param overwrite If true, overwrites existing config file. Defaults to false
   */
  async saveConfigFile(overwrite = false) {
    const configWithoutDefaults: Partial<Config> = {
      ...this.config,
    };

    // Only include non-default values and $schema
    for (const untypedKey in configWithoutDefaults) {
      const key = untypedKey as keyof Config;
      if (
        !["$schema"].includes(key) &&
        configWithoutDefaults[key] === defaultConfig[key]
      ) {
        delete configWithoutDefaults[key];
      }
    }

    const configJSON = JSON.stringify(configWithoutDefaults, null, 2);

    if (this.configPath && !overwrite) {
      throw new Error("Config file already exists");
    }

    if (this.configPath) {
      await writeFile(this.configPath, configJSON);
    } else {
      // Write to the project path
      const packageJSONPath = await findUp("package.json");
      this.projectPath = dirname(packageJSONPath ?? process.cwd());
      this.configPath = join(this.projectPath, CONFIG_FILE_NAME);
      await writeFile(this.configPath, configJSON);
    }
  }

  getConfig(): Config;
  getConfig<K extends keyof Config>(key: K): Config[K];
  getConfig<K extends keyof Config>(key?: K) {
    return key ? this.config?.[key] : this.config;
  }

  getConfigPath() {
    return this.configPath;
  }

  getProjectPath() {
    return this.projectPath ?? process.cwd();
  }

  setConfig(newConfig: Partial<Config>) {
    // Update the config with new values skipping undefined values
    for (const untypedKey in newConfig) {
      const key = untypedKey as keyof Config;
      if (newConfig[key] === undefined) continue;
      (this.config as any)[key] = newConfig[key];
    }
  }
}

export const configStore = new ConfigStore();
