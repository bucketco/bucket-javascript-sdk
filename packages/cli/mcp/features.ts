import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";

import { listFeatures } from "../services/features.js";
import { configStore } from "../stores/config.js";

export function registerMcpFeatures(mcp: McpServer) {
  // Add an addition tool
  mcp.tool("features", {}, async () => {
    const appId = configStore.getConfig("appId");
    if (!appId) {
      return {
        isError: true,
        content: [{ type: "text", text: "Error: appId in config missing" }],
      };
    }
    try {
      const features = await listFeatures(appId);
      return {
        content: [{ type: "text", text: JSON.stringify(features, null, 2) }],
      };
    } catch (error) {
      console.error(error);
      return {
        isError: true,
        content: [{ type: "text", text: `Error: ${error}` }],
      };
    }
  });
}
