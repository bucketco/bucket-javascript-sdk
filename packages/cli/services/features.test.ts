import { describe, it, expect } from "vitest";

import { ConfigFeatureDefs } from "../utils/config.js";
import { genFeatureTypes } from "./features.js";
import { readFile } from "fs/promises";
import path from "path";

describe("genFeatureTypes", () => {
  const features: ConfigFeatureDefs = [
    "feature1",
    { key: "feature2", access: false, config: "string" },
    {
      key: "feature3",
      config: {
        aiModel: "string",
        prompt: {
          text: "string",
          cheekyFactor: "number",
        },
      },
    },
  ];

  it("should generate correct TypeScript output for browser", async () => {
    const outputDir = "test/gen";
    await genFeatureTypes(features, outputDir);
    const dtsOutput = await readFile(
      path.join(outputDir, "_bucket/index.d.ts"),
      "utf-8",
    );
    expect(dtsOutput).toMatchSnapshot("index.d.ts");

    const jsOutput = await readFile(
      path.join(outputDir, "_bucket/index.js"),
      "utf-8",
    );
    expect(jsOutput).toMatchSnapshot("index.d.ts");

    const packageJsonOutput = await readFile(
      path.join(outputDir, "_bucket/package.json"),
      "utf-8",
    );
    expect(packageJsonOutput).toMatchSnapshot("index.d.ts");
  });
});
