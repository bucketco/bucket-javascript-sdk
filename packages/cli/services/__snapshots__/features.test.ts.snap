// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`genFeatureTypes > should generate correct TypeScript output for browser > index.d.ts 1`] = `
"// DO NOT EDIT THIS FILE. IT IS GENERATED BY THE BUCKET CLI AND WILL BE OVERWRITTEN.

export const generatedFeatures: string[];
export interface GeneratedFeatureTypes {
  feature1: {
    key: "feature1", 
    access: true, 
    config: undefined
  }, 
  feature2: {
    key: "feature2", 
    access: false, 
    config: string
  }, 
  feature3: {
    key: "feature3", 
    access: true, 
    config: {
      aiModel: string, 
      prompt: {
        text: string, 
        cheekyFactor: number
      }
    }
  }
}"
`;

exports[`genFeatureTypes > should generate correct TypeScript output for browser > index.d.ts 2`] = `
"// DO NOT EDIT THIS FILE. IT IS GENERATED BY THE BUCKET CLI AND WILL BE OVERWRITTEN.

export const generatedFeatures = [
  {
    key: "feature1", 
    access: true, 
    config: undefined
  }, 
  {
    key: "feature2", 
    access: false, 
    config: "string"
  }, 
  {
    key: "feature3", 
    access: true, 
    config: {
      aiModel: "string", 
      prompt: {
        text: "string", 
        cheekyFactor: "number"
      }
    }
  }
];"
`;

exports[`genFeatureTypes > should generate correct TypeScript output for browser > index.d.ts 3`] = `
"{
  "name": "_bucket",
  "version": "1.0.0",
  "description": "Autogenerated package for typed features for Bucket",
  "main": "index.js",
  "types": "index.d.ts",
  "exports": {
    ".": {
      "import": "./index.js",
      "types": "./index.d.ts"
    }
  }
}"
`;
